graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -248,3 +308,274 @@ def _process_generic_type_definition(generic_type, schema_id, existing_scalars,
+
+    Args:
+        type_name_to_definition: Dict[str, (Interface/Object/Union)TypeDefinition], mapping
+                                 name of types to their definitions
+        scalars: Set[str], names of all scalars in the merged_schema so far
+        cross_schema_edge: CrossSchemaEdge, the edge that we check the validity of
+
+    Raises:
+        - InvalidCrossSchemaEdgeError if the cross schema edge stitches together fields that are
+          not of a scalar type, or stitched together fields that are of different scalar types
+    """
+    field_type_names = []
+
+    for direction, field_reference in (
+        ('out', cross_schema_edge.outbound_side),
+        ('in', cross_schema_edge.inbound_side),
+    ):
+        type_name = field_reference.type_name
+        field_name = field_reference.field_name
+
+        fields = type_name_to_definition[type_name].fields  # List[FieldDefinition]
+        field_type = None
+        for field in fields:
+            if field.name.value == field_name:
+                field_type = field.type
+                while isinstance(field_type, ast_types.NonNullType):  # strip NonNull
+                    field_type = field_type.type
+            break
+
+        if field_type is None:  # should never happen after _check_field_reference_is_valid
+            raise AssertionError(u'Field "{}" unexpectedly not found.'.format(field_name))
+
+        if isinstance(field_type, ast_types.ListType):
+            raise InvalidCrossSchemaEdgeError(
+                u'The {}bound field of cross schema edge "{}" gives a list, while it '
+                u'should be a single scalar'.format(
+                    direction, cross_schema_edge
+                )
+            )
+        elif isinstance(field_type, ast_types.NamedType):
+            if field_type.name.value not in scalars:
+                raise InvalidCrossSchemaEdgeError(
+                    u'The {}bound field of cross schema edge "{}" is of type "{}", which '
+                    u'is not a scalar'.format(
+                        direction, cross_schema_edge, field_type.name.value
+                    )
+                )
+        else:  # since NonNull is stripped, field_type can only be ListType or NamedType
+            raise AssertionError(
+                u'Field has missed type "{}"'.format(type(field_type).__name__)
+            )
+
+        field_type_names.append(field_type.name.value)
+
+    if field_type_names[0] != field_type_names[1]:  # fields return different scalars
+        raise InvalidCrossSchemaEdgeError(
+            u'The outbound and inbound fields of edge "{}" are of different types, '
+            u'"{}" and "{}"; they are expected to be of the same scalar type.'.format(
+                cross_schema_edge, field_type_names[0], field_type_names[1]
+            )
+        )
+
+
+def _add_edge_field(source_type_node, sink_type_node, source_field_name, sink_field_name,
```

Comment:
I might be wrong but it doesn't seem that you are considering inheritance in this function. If an interface has an vertex, then all of the objects that implement it must also have it. If you declare an edge with an union type destination and out_field_only=False, then I would expect that each object in that union also gains a vertex field. Finally, I've never seen union with fields in a GraphQL schema, so I am kind of wary of adding field definitions for them in the AST.

Reply:
Ah interesting, I didn't think very carefully about special treatment for adding an edge to an Interface or a Union. So we don't want to add edges to Union types, and when we add edges to Interface types we want to add that same edge to every concrete Object inheriting the Interface? I'll work on that.

Reply:
@pmantica1 is on vacation, and I believe this has been adequately addressed in the two weeks since this change request, so in the interest of progress I'll dismiss this review and take on responsibility for fixing any follow-up concerns @pmantica1 might have when he returns.